### PR TITLE
MicroSplit config for lightning api draft

### PR DESCRIPTION
**Pull request by CatEek**
## Description

> [!NOTE]  
> **tldr**: 

Lighnting API config fucntion for MS, based on minimal example for HDN. Proposed structure repeats that one of Careamist API, but keeps the model, dataloader and trainer setup separate.

### Background - why do we need this PR?

Extends on recently proposed configuration for HDN (#511 )
**Note: Example below contains CAREAmics style datamodule, which isn't a part of this PR. 
```
config = create_microsplit_configuration(
    experiment_name="ms_test",
    data_type="tiff",
    axes="SYX",
    z_dims=[32] * 4,
    patch_size=(128, 128),
    batch_size=64,
    num_epochs=5,
    predict_logvar="pixelwise",
    train_dataloader_params={"num_workers": 4},
    val_dataloader_params={"num_workers": 4},
    logger=None
)
model = VAEModule(config.algorithm_config)
data_module = create_train_datamodule(
    train_data=train_path,
    val_data=val_path,
    data_type=config.data_config.data_type,
    patch_size=config.data_config.patch_size,
    axes=config.data_config.axes,
    batch_size=config.data_config.batch_size,
    transforms=[],
    train_dataloader_params=config.data_config.train_dataloader_params,
    val_dataloader_params=config.data_config.val_dataloader_params,
)
trainer = Trainer()
trainer.fit()
```
### Overview - what changed?

Reuses configuration function from Careamist API, but creates all modules separately.

## How has this been tested?

Performance haven't been tested

---

**Please ensure your PR meets the following requirements:**

- [x] Code builds and passes tests locally, including doctests
- [x] New tests have been added (for bug fixes/features)
- [ ] Pre-commit passes
- [ ] PR to the documentation exists (for bug fixes / features)

### CODE REVIEW

File: src/careamics/config/algorithms/microsplit_algorithm_model.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from pydantic import ConfigDict
+
+from careamics.config.algorithms.vae_algorithm_model import VAEBasedAlgorithm
+from careamics.config.architectures import LVAEModel
+from careamics.config.loss_model import LVAELossConfig
+
+MICROSPLIT = "MicroSplit"
+
+MICROSPLIT_DESCRIPTION = """MicroSplit is a self-supervised deep learning method for
+microscopy image restoration that combines the strengths of both denoising and
```
Comment (jdeschamps) [inline src/careamics/config/algorithms/microsplit_algorithm_model.py]:
```suggestion
microscopy image splitting that combines the strengths of both denoising and
```

File: src/careamics/config/algorithms/vae_algorithm_model.py [hunk 2]
```diff
@@ -40,14 +40,17 @@ class VAEBasedAlgorithm(BaseModel):
     # defined in SupportedAlgorithm
     # TODO: Use supported Enum classes for typing?
     #   - values can still be passed as strings and they will be cast to Enum
-    algorithm: Literal["musplit", "denoisplit"]
+    algorithm: Literal["hdn", "microsplit", "musplit", "denoisplit"]
```
Comment (jdeschamps) [inline src/careamics/config/algorithms/vae_algorithm_model.py]:
```suggestion
    algorithm: Literal["hdn", "microsplit", "denoisplit"]
```

File: src/careamics/config/algorithms/vae_algorithm_model.py [hunk 3]
```diff
@@ -64,22 +67,25 @@ def algorithm_cross_validation(self: Self) -> Self:
...
                 SupportedLoss.DENOISPLIT,
                 SupportedLoss.DENOISPLIT_MUSPLIT,
             ]:
                 raise ValueError(
-                    f"Algorithm {self.algorithm} only supports loss `denoisplit` "
-                    "or `denoisplit_musplit."
-                )
+                    f"Algorithm {self.algorithm} only supports loss `microsplit`."
+                )  # TODO Update losses configs
+
```
Comment (jdeschamps) [inline src/careamics/config/algorithms/vae_algorithm_model.py]: Here, I am a bit confused because the loss can be any of musplit, denoisplit and denoisplit_musplit, but the error message only says `microsplit`. Which is it?

File: src/careamics/config/configuration_factories.py [hunk 4]
```diff
@@ -1273,3 +1284,575 @@ def create_n2v_configuration(
...
+    model_params["analytical_kl"] = analytical_kl
+
+    return LVAEModel(
+        architecture=SupportedArchitecture.LVAE.value,
+        **model_params,
+    )
+
+
+def _create_vae_based_algorithm(
+    algorithm: Literal["hdn"],
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py]:
```suggestion
    algorithm: Literal["hdn", "microsplit"],
```

File: src/careamics/config/configuration_factories.py [hunk 5]
```diff
@@ -1273,3 +1284,575 @@ def create_n2v_configuration(
...
+    NMLikelihoodConfig | None]
+        The likelihoods and noise models configurations.
+
+    Raises
+    ------
+    ValueError
+        If required parameters are missing for the specified loss_type.
+    """
+    # gaussian likelihood
+    if loss_type in ["musplit", "denoisplit_musplit"]:
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py:1494]: Not a comment to the code, but can we get a summary written somewhere (notion?) of the algorithms (musplit, denoisplit, denoisplit_musplit) and there differences in a table for instance?

It is hard to review otherwise, and we will need to write it on the website anyway!

File: src/careamics/config/configuration_factories.py [hunk 6]
```diff
@@ -1273,3 +1284,575 @@ def create_n2v_configuration(
...
+    encoder_dropout: float = 0.0,
+    decoder_dropout: float = 0.0,
+    nonlinearity: Literal[
+        "None", "Sigmoid", "Softmax", "Tanh", "ReLU", "LeakyReLU", "ELU"
+    ] = "ReLU",
+    analytical_kl: bool = False,
+    predict_logvar: Literal["pixelwise"] | None = None,
+    logvar_lowerbound: Union[float, None] = None,
+    logger: Literal["wandb", "tensorboard", "none"] = "none",
+    model_params: dict | None = None,
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py]: As in https://github.com/CAREamics/careamics/pull/511, is `model_params` of any use? Aren't all model parameters set explicitely here?

File: src/careamics/config/configuration_factories.py [hunk 7]
```diff
@@ -1273,3 +1284,575 @@ def create_n2v_configuration(
...
+
+    return Configuration(
+        experiment_name=experiment_name,
+        algorithm_config=algorithm_params,
+        data_config=data_params,
+        training_config=training_params,
+    )
+
+
+def create_microsplit_configuration(
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py:1694]: Missing loss parameter?

File: src/careamics/config/configuration_factories.py [hunk 8]
```diff
@@ -1273,3 +1284,575 @@ def create_n2v_configuration(
...
+
+    Returns
+    -------
+    Configuration
+        A configuration object for the microsplit algorithm.
+    """
+    transform_list = _list_spatial_augmentations(augmentations)
+
+    loss_config = LVAELossConfig(
+        loss_type="denoisplit_musplit", denoisplit_weight=0.9, musplit_weight=0.1
```
Comment (jdeschamps) [inline src/careamics/config/configuration_factories.py:1775]: If the weights should sum to `1` then I would add a validation in the Pydantic model in this PR.

File: src/careamics/config/loss_model.py [hunk 9]
```diff
@@ -35,7 +35,9 @@ class LVAELossConfig(BaseModel):
         validate_assignment=True, validate_default=True, arbitrary_types_allowed=True
     )
 
-    loss_type: Literal["musplit", "denoisplit", "denoisplit_musplit"]
+    loss_type: Literal[
+        "hdn", "microsplit", "musplit", "denoisplit", "denoisplit_musplit"
```
Comment (jdeschamps) [inline src/careamics/config/loss_model.py:39]:
```suggestion
        "hdn", "microsplit", "denoisplit", "denoisplit_musplit"
```

File: src/careamics/config/support/supported_losses.py [hunk 10]
```diff
@@ -21,8 +21,9 @@ class SupportedLoss(str, BaseEnum):
     MAE = "mae"
     N2V = "n2v"
     # PN2V = "pn2v"
-    # HDN = "hdn"
+    HDN = "hdn"
     MUSPLIT = "musplit"
+    MICROSPLIT = "microsplit"
```
Comment (jdeschamps) [inline src/careamics/config/support/supported_losses.py:26]:
```suggestion
    MICROSPLIT = "microsplit"
```

File: src/careamics/config/support/supported_losses.py [hunk 11]
```diff
@@ -21,8 +21,9 @@ class SupportedLoss(str, BaseEnum):
...
     DENOISPLIT = "denoisplit"
     DENOISPLIT_MUSPLIT = "denoisplit_musplit"
```
Comment (jdeschamps) [inline src/careamics/config/support/supported_losses.py]: I would rename `denoisplit_musplit` into `denoisplit_microsplit` everywhere for consistency. There are a bunch of them laying around.

```suggestion
    DENOISPLIT_MICROSPLIT = "denoisplit_microsplit"
```

File: src/careamics/lightning/lightning_module.py [hunk 12]
```diff
@@ -480,35 +504,55 @@ def predict_step(self, batch: Tensor, batch_idx: Any) -> Any:
...
         else:
-            x = batch
+            x = batch[0] if isinstance(batch, list | tuple) else batch
             aux = []
+            self.model.reset_for_inference(x.shape)
+
+        mmse_list = []
+        for _ in range(self.algorithm_config.mmse_count):
+            # apply test-time augmentation if available
+            if self._trainer.datamodule.prediction_config.tta_transforms:
```
Comment (jdeschamps) [inline src/careamics/lightning/lightning_module.py]: Actually, I think we should remove the test time augmentations for now, they are faulty (users can pass test-time augmentations the model was not trained on) and I am not sure that they are applied correctly.

File: src/careamics/losses/loss_factory.py [hunk 13]
```diff
@@ -59,6 +64,9 @@ def loss_factory(loss: Union[SupportedLoss, str]) -> Callable:
     elif loss == SupportedLoss.MSE:
         return mse_loss
 
+    elif loss == SupportedLoss.HDN:
+        return hdn_loss
+
     elif loss == SupportedLoss.MUSPLIT:
         return musplit_loss
```
Comment (jdeschamps) [inline src/careamics/losses/loss_factory.py]: Shouldn't it be `SupportedLoss.MICROSPLIT`?

**Review by jdeschamps — CHANGES_REQUESTED** [review]
I have not looked at the Lightning module part because I don't think I can understand it, I also skipped the tests.

I have only very minimal suggestions.

File: src/careamics/config/algorithms/vae_algorithm_model.py (hunk 3 above)
Comment (CatEek) [reply src/careamics/config/algorithms/vae_algorithm_model.py]: I think we should drop variations and only leave microsplit

File: src/careamics/config/configuration_factories.py (hunk 5 above)
Comment (CatEek) [reply src/careamics/config/configuration_factories.py:1494]: https://www.notion.so/MicroSplit-Integration-26aa07b1e30c80428cdcde6a9f94bbd2

File: src/careamics/config/loss_model.py (hunk 9 above)
Comment (CatEek) [reply src/careamics/config/loss_model.py:39]: Probably outdated

File: src/careamics/config/support/supported_losses.py (hunk 11 above)
Comment (CatEek) [reply src/careamics/config/support/supported_losses.py]: Should be removed

File: src/careamics/lightning/lightning_module.py (hunk 12 above)
Comment (CatEek) [reply src/careamics/lightning/lightning_module.py]: That's not specifically a problem, but we can discuss separately

File: src/careamics/losses/loss_factory.py (hunk 13 above)
Comment (CatEek) [reply src/careamics/losses/loss_factory.py]: Losses need to be refactored, we should only have microsplit ultimately

File: src/careamics/config/configuration_factories.py (hunk 6 above)
Comment (CatEek) [reply src/careamics/config/configuration_factories.py]: removed

File: src/careamics/config/configuration_factories.py (hunk 7 above)
Comment (CatEek) [reply src/careamics/config/configuration_factories.py:1694]: Shouldn't be there ultimately

File: src/careamics/config/configuration_factories.py (hunk 8 above)
Comment (CatEek) [reply src/careamics/config/configuration_factories.py:1775]: Don't know yet? Added todo

File: src/careamics/config/algorithms/vae_algorithm_model.py (hunk 3 above)
Comment (jdeschamps) [reply src/careamics/config/algorithms/vae_algorithm_model.py]: I agree!

**Review by jdeschamps — APPROVED** [review]
